src/data/polynomial/basic.lean

```diff
@@ -33,77 +35,185 @@ variables {R : Type u} {a b : R} {m n : ℕ}
 section semiring
 variables [semiring R] {p q : polynomial R}
 
-instance : inhabited (polynomial R) := add_monoid_algebra.inhabited _ _
-instance : semiring (polynomial R) := add_monoid_algebra.semiring
+lemma forall_iff_forall_finsupp (P : polynomial R → Prop) :
+  (∀ p, P p) ↔ ∀ (q : add_monoid_algebra R ℕ), P ⟨q⟩ :=
+⟨λ h q, h ⟨q⟩, λ h ⟨p⟩, h p⟩
+
+lemma exists_iff_exists_finsupp (P : polynomial R → Prop) :
+  (∃ p, P p) ↔ ∃ (q : add_monoid_algebra R ℕ), P ⟨q⟩ :=
+⟨λ ⟨⟨p⟩, hp⟩, ⟨p, hp⟩, λ ⟨q, hq⟩, ⟨⟨q⟩, hq⟩ ⟩
+
+/-- The function version of `monomial`. Use `monomial` instead of this one. -/
+@[irreducible] def monomial_fun (n : ℕ) (a : R) : polynomial R := ⟨finsupp.single n a⟩
+@[irreducible] private def add : polynomial R → polynomial R → polynomial R
+| ⟨a⟩ ⟨b⟩ := ⟨a + b⟩
+@[irreducible] private def neg {R : Type u} [ring R] : polynomial R → polynomial R
+| ⟨a⟩ := ⟨-a⟩
```

Comment:
The issue I see with making them semireducible is that, often, we will encounter polynomials that are of the form `polynomial.of_finsupp`: whenever a polynomial will be constructed in terms of monomials (so, in terms of `C` or of `X`), it will be of this form. A middle ground would be to make `0` semireducible (this is essentially the only one for which the lack of defeqness was a little bit painful in the refactor), but leave `add` and `neg` irreducible to make sure that any nontrivial polynomial one defines is not of the form `polynomial.of_finsupp`.

Reply:
I have made `0` semireducible. I don't think making any of the other ones semireducible could help anywhere, but I don't think it would hurt either. So I will change them if you insist.

Reply:
I don't insist.

Reply:
In #15778 `monomial_fun` is made semireducible; only the irreducibility of `add` remains